repairDuplicateVersions.js

```diff
@@ -0,0 +1,447 @@
+        logProgress('finished reading verify log');
+        cb();
+    });
+}
+
+function fetchObjectMetadata(objectUrl, cb) {
+    if (!objectUrl.startsWith('s3://')) {
+        return cb(new Error(`malformed object URL ${objectUrl}: must start with "s3://"`));
+    }
+    const bucketAndObject = objectUrl.slice(5);
+    const url = `http://${BUCKETD_HOSTPORT}/default/bucket/${bucketAndObject}`;
+    return httpRequest('GET', url, null, (err, res) => {
+        if (err) {
+            return cb(err);
+        }
+        if (res.statusCode !== 200) {
+            return cb(new Error(`GET ${url} returned status ${res.statusCode}`));
+        }
+        const md = JSON.parse(res.body);
+        return cb(null, md);
+    });
+}
+
+function putObjectMetadata(objectUrl, objMD, cb) {
+    if (!objectUrl.startsWith('s3://')) {
+        return cb(new Error(`malformed object URL ${objectUrl}: must start with "s3://"`));
+    }
+    const bucketAndObject = objectUrl.slice(5);
+    const url = `http://${BUCKETD_HOSTPORT}/default/bucket/${bucketAndObject}`;
+    return httpRequest('POST', url, JSON.stringify(objMD), (err, res) => {
+        if (err) {
+            return cb(err);
+        }
+        if (res.statusCode !== 200) {
+            return cb(new Error(`POST ${url} returned status ${res.statusCode}`));
+        }
+        return cb();
+    });
+}
+
+function genSproxydKey(fromKey) {
+    // See sproxydclient:lib/keygen.js for details on how sproxyd keys
+    // are generated.
+    //
+    // Here, instead of needing the original info to construct the
+    // key, we reuse the fields from the existing key and regenerate
+    // the random parts.
+
+    const rand = crypto.randomBytes(11);
+    return [rand.slice(0, 8).toString('hex').toUpperCase(),
+            fromKey.slice(16, 32),
+            rand.slice(8, 11).toString('hex').toUpperCase(),
+            fromKey.slice(38, 40)].join('');
+}
+
+function copySproxydKey(objectUrl, sproxydKey, cb) {
+    const cbOnce = jsutil.once((err, newKey) => {
+        if (err) {
+            status.sproxydKeysCopyErrors += 1;
+        } else {
+            status.sproxydKeysCopied += 1;
+        }
+        cb(err, newKey);
+    });
+    const newKey = genSproxydKey(sproxydKey);
+    const sproxydSourceUrl =
+          new URL(`http://${SPROXYD_HOSTPORT}/${sproxydAlias}/${sproxydKey}`);
+    const sproxydDestUrl =
+          new URL(`http://${SPROXYD_HOSTPORT}/${sproxydAlias}/${newKey}`);
+    const sourceReq = http.request({
+        hostname: sproxydSourceUrl.hostname,
+        port: sproxydSourceUrl.port,
+        path: sproxydSourceUrl.pathname,
+        method: 'GET',
+        agent: httpAgent,
+    }, sourceRes => {
+        if (sourceRes.statusCode !== 200) {
+            log.error('sproxyd returned HTTP error code', {
+                objectUrl,
+                sproxydKey,
+                httpCode: sourceRes.statusCode,
+            });
+            sourceRes.on('data', () => {});
```

Comment:
is this required?

Reply:
I'm not sure but it could likely be replaced by a `sourceRes.resume()` call instead, so that the data is consumed. I think we need to consume the data explicitly though because it's not used elsewhere.

Reply:
Re-reading it, my bad, I indeed think we should get rid of it because we're doing `sourceReq.pipe(targetReq)` below.

Reply:
Ah no, looking even more closely, this is the error case, so we're not reading the response otherwise and need to consume it.